src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs

```diff
@@ -123,9 +216,73 @@ public RegexCode RegexCodeFromRegexTree(RegexTree tree)
+                    }
+                    if (curNode.Type == RegexNode.Alternate)
+                    {
+                        curNode.cumulativeNumberOfChildren += curNode.Child(curChild).cumulativeNumberOfChildren;
+                        // Because the current node is an alternate node, we expand our words.
+                        if (curChild == curNode.ChildCount() - 1)
+                        {
+                            int numberOfCurrentWords = words.Count - curNode.cumulativeNumberOfChildren;
+                            if (numberOfCurrentWords > 0)
+                            {
+                                List<string> newWords = new List<string>(numberOfCurrentWords * curNode.cumulativeNumberOfChildren);
+                                for (int i = 0; i < numberOfCurrentWords; i++)
+                                {
+                                    string concatenateWord = words[i];
+                                    for (int j = numberOfCurrentWords; j < words.Count; j++)
+                                    {
+                                        newWords.Add(concatenateWord + words[j]);
+                                    }
+                                }
+                                words = newWords;
+                                curNode.cumulativeNumberOfChildren = words.Count;
+                            }
+
+                        }
+                    }
+                    if (curNode.Type == RegexNode.Concatenate)
+                    {
+                        RegexNode childNode = curNode.Child(curChild);
+                        if (EffectiveChildIsAnAlternateNode(curNode, curChild))
+                        {
+                            curNode.cumulativeNumberOfChildren = words.Count;
+                        }
+                        else if (EffectiveChildIsAMultiNode(curNode, curChild, out RegexNode multiNode))
+                        {
+                            string multiStr = multiNode.Str!;
+                            int numberOfCurrentWords = words.Count - multiNode.cumulativeNumberOfChildren;
+                            if (numberOfCurrentWords > 0)
+                            {
+                                for (int i = 0; i < numberOfCurrentWords; i++)
+                                {
+                                    words[i] += multiStr;
+                                }
+                                for (int i = words.Count - 1; i >= numberOfCurrentWords; i--)
+                                {
+                                    words.RemoveAt(i);
+                                }
+                            }
+                            curNode.cumulativeNumberOfChildren = words.Count;
+                        }
+                    }
+                }
                 curChild++;
             }
 
+            if (useAhoCorasick)
+            {
+                ahoCorasickWords = words;
```

Comment:
If someone wants to test locally, putting a breakpoint here and inspecting `words` can tell you if the words list was populated correctly